machine_learning/neural_network.cpp

```diff
@@ -0,0 +1,719 @@
+                    /*
+                        Format in which model is saved:
+
+                        total_layers
+                        neurons(1st Layer) activation_name(1st Layer)
+                        kernal_shape(1st Layer)
+                        kernal_values
+                        .
+                        .
+                        .
+                        neurons(Nth Layer) activation_name(Nth Layer)
+                        kernal_shape(Nth Layer)
+                        kernal_value
+
+                        For Example, pretrained model with 3 layers:
+
+                        3
+                        4 none
+                        4 4
+                        1 0 0 0 
+                        0 1 0 0 
+                        0 0 1 0 
+                        0 0 0 1 
+                        6 relu
+                        4 6
+                        -1.88963 -3.61165 1.30757 -0.443906 -2.41039 -2.69653 
+                        -0.684753 0.0891452 0.795294 -2.39619 2.73377 0.318202 
+                        -2.91451 -4.43249 -0.804187 2.51995 -6.97524 -1.07049 
+                        -0.571531 -1.81689 -1.24485 1.92264 -2.81322 1.01741 
+                        3 sigmoid
+                        6 3
+                        0.390267 -0.391703 -0.0989607 
+                        0.499234 -0.564539 -0.28097 
+                        0.553386 -0.153974 -1.92493 
+                        -2.01336 -0.0219682 1.44145 
+                        1.72853 -0.465264 -0.705373 
+                        -0.908409 -0.740547 0.376416 
+
+                    */
+                    // Saving model in the same format
+                    out_file << layers.size();
+                    out_file << std::endl;
+                    for(const auto &layer : this -> layers) {
+                        out_file << layer.neurons << ' ' << layer.activation << std::endl;
+                        const auto shape = get_shape(layer.kernal);
+                        out_file << shape.first << ' ' << shape.second << std::endl;
+                        for(const auto &row : layer.kernal) {
+                            for(const auto &val : row) {
+                                out_file << val << ' ';
+                            }
+                            out_file << std::endl;
+                        }
+                    }
+                    std::cout << "INFO: Model saved successfully with name : ";
+                    std::cout << file_name << std::endl;
+                    return;
+                }
+
+                /**
+                 * Function to load earlier saved model.
+                 * @param file_name file from which model will be loaded (*.model)
+                 * @return instance of NeuralNetwork class with pretrained weights
+                 */
+                NeuralNetwork load_model (const std::string &file_name) {
```

Comment:
Since this function returns a class instance, for it to be usable, it'd require the implementation of a `copy constructor` for this class.

Reply:
I have tested it with default ```copy constructor``` and it works fine.